content/blog/how-tikv-accesses-data.md

```diff
@@ -0,0 +1,263 @@
+---
+title: How TiKV reads and writes
+date: 2019-12-06
+author: Tang Liu
+---
+
+This article introduces in detail how TiKV handles read and write operations. After reading this document, you will learn how TiKV, as a distributed database, stores the data contained in a write request and how it retrieves the corresponding data with consistency guaranteed.
+
+## Before you read it
+
+Before we begin, we need to introduce some essential concepts of TiKV to help you fully understand the process. If you are already familiar with these, you can skip to the next section.
+
+### Raft
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/raft-process-in-tikv.png"
+    caption="Raft Process in TiKV"
+    number="" >}}
+
+TiKV uses the Raft consensus algorithm to ensure data safety and consistency. By default, three replicas are supported, and these three replicas form a Raft Group.
+
+When the client needs to write some data, it sends the request to the Raft Leader. This process is called **Propose** in TiKV. The Leader encodes the operation into an entry and writes it into its own **Raft Log**. This is called **Append**.
+
+The Leader will also copy the entry to other Followers through the Raft algorithm. This is called **Replicate**. After the Follower receives the entry, it will also perform an **Append** operation, and in the meantime inform the Leader that the Append is successful.
+
+When the Leader finds that the entry has been appended by majority of nodes, it considers that the entry Committed. Then, it can decode the operations in the entry, execute them, and apply them to the state machine. This is called **Apply**.
+
+TiKV supports a feature called Lease Read. For Read requests, they can be sent directly to the Leader. If the Leader determines that its lease has not expired, it directly provides the Read service without having to go through the Raft process. If the lease has expired, the Leader will be forced to go through the Raft process to renew the lease before it provides the Read service.
+
+
+### Multi Raft
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/multiraft.png"
+    caption="Multi Raft"
+    number="" >}}
+
+Because one Raft Group only processes a limited amount of data, we split the data into multiple Raft Groups, each of which correspond to a Region. The way of splitting is to slice by range. That is, we sort the keys of the data in byte order to form an infinite sorted map, and then slice it into a segments of continuous key range. Each key range is treated as a Region. The range of the Region uses the front-end and back-open mode [start, end). The end key of the previous region is the start key of the next region.
```

Comment:
I suggest we use math formula here. :) You can check the hugo preview.

```suggestion
Because one Raft Group only processes a limited amount of data, we split the data into multiple Raft Groups, each of which correspond to a Region. The way of splitting is to slice by range. That is, we sort the keys of the data in byte order to form an infinite sorted map, and then slice it into a segments of continuous key range. Each key range is treated as a Region. The range of the Region uses the front-end and back-open mode \\([start, end)\\). The end key of the previous region is the start key of the next region.
```